File: sdk/swift/Sources/ConduytKit/Modules/Encoder.swift
import Foundation

/// Typed wrapper for the CONDUYT quadrature encoder module.
@available(iOS 15.0, macOS 12.0, *)
public class ConduytEncoder {
    private let device: ConduytDevice
    private let moduleID: UInt8

    public init(device: ConduytDevice, moduleID: UInt8) {
        self.device = device
        self.moduleID = moduleID
    }

    /// Claim two pins for A/B inputs.
    public func attach(pinA: UInt8, pinB: UInt8) async throws {
        _ = try await device.modCmd(Data([moduleID, 0x01, pinA, pinB]))
    }

    /// Read the current accumulated tick count. Firmware echoes module id, then a LE int32.
    public func read() async throws -> Int32 {
        let resp = try await device.modCmd(Data([moduleID, 0x02]))
        guard resp.count >= 5 else {
            throw ConduytDeviceError.nak(code: 0xFE, name: "SHORT_RESPONSE")
        }
        return resp.readLE(int32At: 1)
    }

    /// Reset the count to zero.
    public func reset() async throws {
        _ = try await device.modCmd(Data([moduleID, 0x03]))
    }
}

File: sdk/rust/src/modules/encoder.rs
//! CONDUYT Quadrature Encoder Module — typed wrapper

use crate::device::{Device, DeviceError};
use crate::transports::Transport;

/// Encoder module wrapper.
pub struct Encoder<'a, T: Transport> {
    device: &'a mut Device<T>,
    module_id: u8,
}

impl<'a, T: Transport> Encoder<'a, T> {
    pub fn new(device: &'a mut Device<T>, module_id: u8) -> Self {
        Self { device, module_id }
    }

    /// Claim two pins for A/B inputs.
    pub fn attach(&mut self, pin_a: u8, pin_b: u8) -> Result<(), DeviceError<T::Error>> {
        self.device.mod_cmd(&[self.module_id, 0x01, pin_a, pin_b])?;
        Ok(())
    }

    /// Read the accumulated tick count. Firmware echoes module id, then a LE int32.
    pub fn read(&mut self) -> Result<i32, DeviceError<T::Error>> {
        let resp = self.device.mod_cmd(&[self.module_id, 0x02])?;
        if resp.len() < 5 {
            return Err(DeviceError::WireError(crate::wire::WireError::IncompletePkt));
        }
        Ok(i32::from_le_bytes([resp[1], resp[2], resp[3], resp[4]]))
    }

    /// Reset the count to zero.
    pub fn reset(&mut self) -> Result<(), DeviceError<T::Error>> {
        self.device.mod_cmd(&[self.module_id, 0x03])?;
        Ok(())
    }
}
